### PR TITLE
Don't share host_array when receiving from network

### PR DESCRIPTION
**Pull request by crusaderky**
- This PR is blocked by and incorporates https://github.com/dask/distributed/pull/8312

Resolve issues with memory deallocation:

1. Two or more numpy or pandas objects are packed into the same network message by `WorkerStateMachine._select_for_gather`, `scatter`, or `Client.gather`. After they are received, one of the objects is dereferenced, but its memory won't be released until *all* objects with a buffer in the original message have been dereferenced.

2. An object with both buffers and non-trivial amounts of pure-pickle data - such as a pandas.DataFrame with object columns - is sent over the network. For as long as the object lives, the memory holding the pickled version of the object column won't be released.

3. In #8282, when using a MemoryBuffer the shards that have already been merged into output chunks are not dereferenced until *all* shards on the same worker have been merged. This is because shards belonging to different output chunks were sent over within the same RPC call.

## Notes
- These issues only apply to uncompressed data.
- Use case 2 also afflicts the SpillBuffer. It is out of scope of this PR.

### CODE REVIEW

**Comment by crusaderky** [comment]
A/B tests results:
- `test_array.py::test_basic_sum[fast-thin]` is 15% slower
- `test_array.py::test_quadratic_mean` is 5% slower
- `test_array.py::test_vorticity` is 10% slower
- `test_dataframe.py::test_dataframe_align` uses 15% less memory
- `test_join.py::test_set_index_on_uber` is 10% faster and uses 10% less memory

I'm going to rework this and revert to the previous behaviour when there's only one buffer and the non-buffer data is less than 5% of the total.

**Comment by crusaderky** [comment]
A/B test results:

- no wall time changes whatsoever
- test_dataframe_align uses 10% less avg memory and 15% less peak memory
- test_set_index_on_uber_lyft[tasks] uses 10% less avg memory and 15% less peak memory

image
image

**Comment by crusaderky** [comment]
This PR is blocked by and incorporates https://github.com/dask/distributed/pull/8312
Ready for review.

File: distributed/comm/tcp.py [hunk 1]
```diff
@@ -371,6 +354,79 @@ def extra_info(self):
...
+    buf = host_array(n)
+
+    for i, j in sliding_window(
+        2,
+        range(0, n + OPENSSL_MAX_CHUNKSIZE, OPENSSL_MAX_CHUNKSIZE),
+    ):
+        chunk = buf[i:j]
+        chunk_nbytes = chunk.nbytes
+        n = await stream.read_into(chunk)  # type: ignore[arg-type]
+        assert n == chunk_nbytes, (n, chunk_nbytes)
```
Comment (milesgranger) [inline distributed/comm/tcp.py:370]: Apologies for any naive things here, for my own understanding, this assert is a sanity check, as [according to the docs for read_into](https://www.tornadoweb.org/en/stable/iostream.html#tornado.iostream.BaseIOStream.read_into) this won't return until `chunk` (of which is `chunk_nbytes` in length) is completely filled. Is there a potential that it won't be filled due to chunk being larger than any remaining bytes, and thus sit idle?

I would suppose not judging by the caller setting `n` but got slightly confused with the openssl sizes and that may make a chunk larger than any remaining stream data.

Comment (milesgranger) [reply distributed/comm/tcp.py:370]: Or now that I look at it again.. bytes in the stream would always be larger than the buffer, and thus stream would always have bytes to write given we're accounting for OpenSSL sizing. Do I have that right?

File: distributed/comm/tcp.py [hunk 2]
```diff
@@ -371,6 +354,79 @@ def extra_info(self):
...
+
+    header = pack_frames_prelude(frames)
+    header = struct.pack("Q", nbytes(header) + frames_nbytes_nosplit) + header
+    header_nbytes = nbytes(header)
+
+    frames = [header, *frames]
+    frames_nbytes = [header_nbytes, *frames_nbytes]
+    frames_nbytes_total += header_nbytes
+
+    if frames_nbytes_total < 2**17:  # 128kiB
```
Comment (fjetter) [inline distributed/comm/tcp.py:422]: Given the testing you did recently. Do you think this number still makes sense? Something to look into or not worth the effort?

File: distributed/comm/tests/test_comms.py [hunk 3]
```diff
@@ -1379,3 +1380,95 @@ def test_register_backend_entrypoint(tmp_path):
...
+    assert get_host_array(a) is not get_host_array(b)
+
+
+@pytest.mark.parametrize(
+    "nbytes_np,nbytes_other,expect_separate_buffer",
+    [
+        (1, 0, False),  # <2 kiB (including prologue and msgpack header)
+        (1, 1800, False),  # <2 kiB
+        (1, 2100, True),  # >2 kiB
+        (200_000, 9500, False),  # <5% of numpy array
```
Comment (fjetter) [inline distributed/comm/tests/test_comms.py:1433]: IIUC the 9500 extra bytes here will be written to the same memory buffer the numpy array is using, i.e. those 9500 bytes will only be released once the numpy array is released. Similarly, the numpy array will only be released once the other thing has been released. However, that _other thing_ is guaranteed to be a bytes object or some header information or some other garbage that is guaranteed to be released after the message is deserialized.

So, in other words, we're accepting a memory overhead of up to 5% for numpy arrays/arrow tables/etc. (and previously this could've been a multiple, depending on how large a single fetch was)

**Review by fjetter — APPROVED** [review]
Changes overall LGTM. Most/all of my comments are rather for clarification but as long as I didn't entirely misunderstand this, I'm happy to merge.

File: distributed/comm/tcp.py (hunk 1 above)
Comment (crusaderky) [reply distributed/comm/tcp.py:370]: yes that assertion is just defensive programming.

File: distributed/comm/tcp.py [hunk 4]
```diff
@@ -220,18 +220,16 @@ async def read(self, deserializers=None):
...
-            frames = host_array(frames_nbytes)
-            for i, j in sliding_window(
-                2,
-                range(0, frames_nbytes + OPENSSL_MAX_CHUNKSIZE, OPENSSL_MAX_CHUNKSIZE),
-            ):
-                chunk = frames[i:j]
-                chunk_nbytes = chunk.nbytes
-                n = await stream.read_into(chunk)
-                assert n == chunk_nbytes, (n, chunk_nbytes)
+            # Don't store multiple numpy or parquet buffers into the same buffer, or
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/comm/tcp.py:223]: wrong location for the comment?
parquet was a typo; i meant arrow.

File: distributed/comm/tcp.py (hunk 2 above)
Comment (crusaderky) [reply distributed/comm/tcp.py:422]: It looks about right given my recent testing.

File: distributed/comm/tests/test_comms.py (hunk 3 above)
Comment (crusaderky) [reply distributed/comm/tests/test_comms.py:1433]: > So, in other words, we're accepting a memory overhead of up to 5% for numpy arrays/arrow tables/etc.

This is correct. This is really only material for pandas objects with substantial pure-python index / columns / other metadata; numpy objects tend to be <100 bytes worth of metadata.

> (and previously this could've been a multiple, depending on how large a single fetch was)

It was worse than a multiple.
There were two nightmare scenarios:
- pandas dataframe heavy with object string columns, with some numerical columns. The whole serialized data for the  object columns remains alive for as long as the deserialized object is alive, because it's referenced by the numerical columns.
- The key at the top of the `WorkerStateMachine.fetch` heap is a 49 MiB nump array. The second object in the heap from the same worker is a 1 MiB numpy array (or vice versa). The two are fetched together (`distributed.worker.transfer.message-bytes-limit: 50 MiB`). The 49 MiB array will survive its own `free-keys` command, as it is referenced by the 1 MiB array.
